File: algorithms/0101.symmetric-tree/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match(root) {
            Some(node) => {
                let left = node.borrow().left.clone();
                let right = node.borrow().right.clone();
                Self::is_mirror(left, right)
            },
            None => true,
        }
    }

    fn is_mirror(a: Option<Rc<RefCell<TreeNode>>>, b: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match(a, b) {
            (Some(a), Some(b)) => {
                let a = a.borrow();
                let b = b.borrow();

                if a.val == b.val {
                    Self::is_mirror(a.left.clone(), b.right.clone())
                    && Self::is_mirror(b.left.clone(), a.right.clone())
                } else {
                    false
                }
            },
            (None, None) => true,
            _ => false,
        }
    }
}

File: algorithms/0101.symmetric-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	return root == nil || is_mirror(root.Left, root.Right)
}

func is_mirror(a *TreeNode, b *TreeNode) bool {
	if a == nil || b == nil {
		return a == nil && b == nil
	}

	if a.Val == b.Val {
		return is_mirror(a.Left, b.Right) && is_mirror(b.Left, a.Right)
	} else {
		return false
	}
}
